numpy/testing/_private/utils.py

```diff
@@ -353,13 +353,13 @@ def assert_equal(actual, desired, err_msg='', verbose=True, *, strict=False):
         for k, i in desired.items():
             if k not in actual:
                 raise AssertionError(repr(k))
-            assert_equal(actual[k], desired[k], f'key={k!r}\n{err_msg}',
+            assert_equal(actual[k], i, f'key={k!r}\n{err_msg}',
                          verbose)
         return
     if isinstance(desired, (list, tuple)) and isinstance(actual, (list, tuple)):
         assert_equal(len(actual), len(desired), err_msg, verbose)
-        for k in range(len(desired)):
-            assert_equal(actual[k], desired[k], f'item={k!r}\n{err_msg}',
+        for k, i in enumerate(desired):
+            assert_equal(actual[k], i, f'item={k!r}\n{err_msg}',
```

Comment:
This is less readable than before, but maybe we have to accept these changes to have the benefits of ruff. Renaming the `i` would help here though. Another option would be to turn of the performance rules for the tests (as test performance is typically not critical)

Reply:
It's very much out-of-scope here, but generally speaking it's preferable to use parametric tests instead of putting assertions in a for-loop 🤷🏻 

But even so, I agree that it wouldn't hurt to rename `i` to something more descriptive.

Reply:
To tell the truth, the `range(len))` → `enumerate()` change has **not** been suggested by _ruff_. I added it for consistency with the `desired.keys()` → `desired.items()` change a few lines before.

I am happy to revert the `range(len))` → `enumerate()` change and disable the `desired.keys()` → `desired.items()`, globally or locally.

Note that PERF rules are not necessarily about "performance" as such. In most cases, these micro-optimisations will have only (very) marginal effect on actual global performance. Rather, the idea of all these rules is to use consistent Python style.